cachi2/core/package_managers/npm.py

```diff
@@ -173,6 +185,115 @@ def get_sbom_components(self) -> list[dict[str, str]]:
         packages = self._dependencies if self.lockfile_version == 1 else self._packages
         return [{"name": package.name, "version": package.version} for package in packages]
 
+    def get_dependencies_to_download(self) -> Dict[str, Dict[str, Any]]:
+        """Return a list of URL dependencies to download."""
+        packages = self._dependencies if self.lockfile_version == 1 else self._packages
+        return {
+            package.name: {
+                "version": package.version,
+                "url": package.resolved_url,
+                "integrity": package.integrity,
+            }
+            for package in packages
+            if "file:" not in package.resolved_url
+        }
+
+
+def _get_external_requirement_filepath(
+    download_path: RootedPath, package: str, requirement_url: str
+) -> Path:
+    if requirement_url.startswith("https://"):
+        algorithm = "sha512"
+        digest = "YOLO"
+        # TODO: Do I have to download that package first to get algorithm and digest???
+        filepath = Path(
+            download_path, f"external-{package}", f"{package}-external-{algorithm}-{digest}.tgz"
+        )
+    elif (
+        requirement_url.startswith("git+ssh://")
+        or requirement_url.startswith("github://")
+        or requirement_url.startswith("git+https://")
+    ):
+        parsed_url = urlparse(requirement_url)
+        namespace = parsed_url.path.removeprefix("/").split("/")[0]
+        repo = parsed_url.path.removeprefix("/").split("/")[-1].removesuffix(".git")
+        ref = parsed_url.fragment
+
+        filepath = Path(
+            download_path,
+            parsed_url.netloc,  # host
+            namespace,
+            repo,
+            f"{repo}-external-gitcommit-{ref}.tgz",
+        )
+    else:
+        raise ValueError(f"{requirement_url} is neither 'url' nor 'vcs'")
+    return filepath
+
+
+def clone_repo_pack_archive(
+    url: str,
+    download_path: Path,
+) -> None:
+    """
+    Clone a repository and pack its content as tar.
+
+    :param str url: URL for file download
+    :param str download_path: Path to download file to
+    :raise FetchError: If download failed
+    """
+    parse_url = url.split("#")
+    repo_url = parse_url[0]
+    repo_commit = parse_url[1] if len(parse_url) > 1 else None
+
+    # Create missing directories
+    directory = os.path.dirname(download_path)
+    os.makedirs(directory, exist_ok=True)
+
+    try:
+        clone_as_tarball(str(repo_url), str(repo_commit), download_path)
+    except FetchError:
+        log.info(f"Try to download git source with https:// {url}")
+        clone_as_tarball(
```

Comment:
How do we want to deal with fallback function to second try to retrieve source not via ssh but https://?? Logs contain error now. http://pastebin.test.redhat.com/1100644

Reply:
Side note / future improvement:

NPM uses git-clone only if it doesn't know how to get the tarball without cloning. For github, gitlab, bitbucket and others, npm doesn't actually git-clone but gets the tarball directly from a url, e.g. https://github.com/npm/hosted-git-info/blob/5caeb24a4c96e54446903ecf8caad0f9175f812f/lib/hosts.js#L50